Sqlaner Error "NULL_TYPE is illegal in thrift stage"

<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
1. prepare data
```sql
DROP DATABASE IF EXISTS d7;
CREATE DATABASE d7;
USE d7;
admin set frontend config ("enable_decimal_v3" = "true");
CREATE TABLE t0 (c_0_0 DATETIME NOT NULL ,c_0_1 SMALLINT MAX NULL ,c_0_2 CHAR(21) MIN NULL ,c_0_3 SMALLINT REPLACE NULL ,c_0_4 DOUBLE MIN NULL ) AGGREGATE KEY (c_0_0) PARTITION BY RANGE(c_0_0) (START ("2010-01-01") END ("2022-12-31") EVERY (INTERVAL 115 day)) DISTRIBUTED BY HASH (c_0_0) BUCKETS 3 properties("replication_num"="3") ;
CREATE TABLE IF NOT EXISTS t1 (c_1_0 DATETIME NULL ,c_1_1 DATETIME NULL ,c_1_2 DECIMAL(14, 2)  NULL ,c_1_3 CHAR(1) NOT NULL ,c_1_4 DECIMAL(1, 0)  NOT NULL ,c_1_5 BOOLEAN NOT NULL ,c_1_6 BOOLEAN NOT NULL ,c_1_7 DATE NOT NULL ,c_1_8 INT NOT NULL ,c_1_9 BOOLEAN NOT NULL ,c_1_10 SMALLINT NOT NULL ,c_1_11 DECIMAL(1, 0)  NULL ,c_1_12 DECIMAL(22, 9)  NULL , INDEX index1_c_1_7(`c_1_7`) USING BITMAP) UNIQUE KEY (c_1_0,c_1_1,c_1_2,c_1_3,c_1_4,c_1_5,c_1_6,c_1_7,c_1_8,c_1_9,c_1_10,c_1_11) PARTITION BY RANGE(c_1_1) (START ("2010-01-01") END ("2022-12-31") EVERY (INTERVAL 189 day)) DISTRIBUTED BY HASH (c_1_11,c_1_8,c_1_10,c_1_0,c_1_6,c_1_5,c_1_1,c_1_2) BUCKETS 3 properties("replication_num"="3", "bloom_filter_columns"="c_1_3") ;
INSERT INTO t1 VALUES ('2010-01-01 08:00:14', '2010-01-26 08:00:14', NULL, '', 2, true, false, '2010-01-19', 20, false, 3, 8, 6192.758257631), ('2010-01-19 08:00:14', '2010-01-10 08:00:14', 5319117516.63, '', 5, true, false, '2010-01-19', 19, true, 1, 1, 971.393309088);
INSERT INTO t0 VALUES ('2010-01-27 08:00:14', 1, 'ZZq8vLerj', 1, 0.2936273522796976);
DELETE FROM t0 WHERE c_0_0='1969-12-20 06:48:25';
INSERT INTO t0 VALUES ('2010-01-12 08:00:14', 15, 'b', 12, 0.2936273522796976), ('2010-01-27 08:00:14', 4, '', 2, 0.841342676952513);
INSERT INTO t0 VALUES ('2010-01-29 08:00:14', 12, 'AW2sGAI', 13, 0.9733227055445478), ('2010-01-18 08:00:14', 14, 'A', 14, -1.041094952E9);
INSERT INTO t1 VALUES ('2010-01-23 08:00:14', '2010-01-20 08:00:14', 378554168.22, '', 0, true, true, '2010-01-05', 23, false, 8, 0, 4204.242449469);
INSERT INTO t1 VALUES ('2010-01-18 08:00:14', '2010-01-31 08:00:14', 2228380427.28, '', 8, false, true, '2010-01-11', 29, true, 8, 6, 3073.473382929);
INSERT INTO t0 VALUES ('2010-01-26 08:00:14', 1, '4n8', 6, NULL), ('2010-01-27 08:00:14', 8, NULL, 2, 0.9430070523121846);
INSERT INTO t1 VALUES ('2010-01-15 08:00:14', '2010-01-30 08:00:14', 6495038508.11, '', 1, true, false, '2010-01-30', 8, true, 10, 4, 5796.051379658);
INSERT INTO t1 VALUES ('2009-12-31 08:00:14', '2010-01-31 08:00:14', 4578845300.42, '', 1, false, false, '2010-01-10', 20, false, 9, 6, 4400.950067720), ('2010-01-29 08:00:14', '2010-01-11 08:00:14', NULL, '', 1, true, false, '2010-01-07', 1, true, 13, 8, 6263.032768170), ('2010-01-20 08:00:14', '2010-01-13 08:00:14', NULL, '', 0, true, false, '2010-01-29', 4, true, 9, 2, 615.177587787), ('2010-01-19 08:00:14', NULL, 3076309094.25, '', 1, false, false, '2010-01-14', 8, true, 14, 5, 9162.715913154);
```
2. query
```sql
SELECT DISTINCT * FROM (SELECT DISTINCT NULL FROM  t1 AS t1_88 FULL JOIN (SELECT t0_87.c_0_0, t0_87.c_0_1, t0_87.c_0_2, t0_87.c_0_3, t0_87.c_0_4 FROM t0 AS t0_87) subt0 ON t1_88.c_1_5 = subt0.c_0_2 AND CASE  WHEN false THEN t1_88.c_1_5  WHEN true THEN t1_88.c_1_5 ELSE t1_88.c_1_5 END < subt0.c_0_2 WHERE NULL UNION ALL SELECT DISTINCT NULL FROM  t1 AS t1_88 FULL JOIN (SELECT t0_87.c_0_0, t0_87.c_0_1, t0_87.c_0_2, t0_87.c_0_3, t0_87.c_0_4 FROM t0 AS t0_87) subt0 ON t1_88.c_1_5 = subt0.c_0_2 AND CASE  WHEN false THEN t1_88.c_1_5  WHEN true THEN t1_88.c_1_5 ELSE t1_88.c_1_5 END < subt0.c_0_2 WHERE (NOT (NULL)) UNION ALL SELECT DISTINCT NULL FROM  t1 AS t1_88 FULL JOIN (SELECT t0_87.c_0_0, t0_87.c_0_1, t0_87.c_0_2, t0_87.c_0_3, t0_87.c_0_4 FROM t0 AS t0_87) subt0 ON t1_88.c_1_5 = subt0.c_0_2 AND CASE  WHEN false THEN t1_88.c_1_5  WHEN true THEN t1_88.c_1_5 ELSE t1_88.c_1_5 END < subt0.c_0_2 WHERE ((NULL) IS NULL)) t;

```

### Expected behavior (Required)
NO ERROR

### Real behavior (Required)
```
--java.lang.AssertionError: SELECT DISTINCT * FROM (SELECT DISTINCT NULL FROM  t1 AS t1_88 FULL JOIN (SELECT t0_87.c_0_0, t0_87.c_0_1, t0_87.c_0_2, t0_87.c_0_3, t0_87.c_0_4 FROM t0 AS t0_87) subt0 ON t1_88.c_1_5 = subt0.c_0_2 AND CASE  WHEN false THEN t1_88.c_1_5  WHEN true THEN t1_88.c_1_5 ELSE t1_88.c_1_5 END < subt0.c_0_2 WHERE NULL UNION ALL SELECT DISTINCT NULL FROM  t1 AS t1_88 FULL JOIN (SELECT t0_87.c_0_0, t0_87.c_0_1, t0_87.c_0_2, t0_87.c_0_3, t0_87.c_0_4 FROM t0 AS t0_87) subt0 ON t1_88.c_1_5 = subt0.c_0_2 AND CASE  WHEN false THEN t1_88.c_1_5  WHEN true THEN t1_88.c_1_5 ELSE t1_88.c_1_5 END < subt0.c_0_2 WHERE (NOT (NULL)) UNION ALL SELECT DISTINCT NULL FROM  t1 AS t1_88 FULL JOIN (SELECT t0_87.c_0_0, t0_87.c_0_1, t0_87.c_0_2, t0_87.c_0_3, t0_87.c_0_4 FROM t0 AS t0_87) subt0 ON t1_88.c_1_5 = subt0.c_0_2 AND CASE  WHEN false THEN t1_88.c_1_5  WHEN true THEN t1_88.c_1_5 ELSE t1_88.c_1_5 END < subt0.c_0_2 WHERE ((NULL) IS NULL)) t;
--	at sqlancer.common.query.QueryAdapter.checkException(QueryAdapter.java:98)
--	at sqlancer.common.query.QueryAdapter.executeAndGet(QueryAdapter.java:128)
--	at sqlancer.ComparatorHelper.getResultSetFirstColumnAsString(ComparatorHelper.java:54)
--	at sqlancer.ComparatorHelper.getCombinedResultSetNoDuplicates(ComparatorHelper.java:164)
--	at sqlancer.starrocks.oracle.tlp.StarRocksTLPDistinctOracle.check(StarRocksTLPDistinctOracle.java:43)
--	at sqlancer.ProviderAdapter.generateAndTestDatabase(ProviderAdapter.java:55)
--	at sqlancer.Main$DBMSExecutor.run(Main.java:326)
--	at sqlancer.Main$2.run(Main.java:521)
--	at sqlancer.Main$2.runThread(Main.java:498)
--	at sqlancer.Main$2.run(Main.java:488)
--	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
--	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
--	at java.lang.Thread.run(Thread.java:748)
--Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: NULL_TYPE is illegal in thrift stage
--	at sun.reflect.GeneratedConstructorAccessor5.newInstance(Unknown Source)
--	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
--	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
--	at com.mysql.jdbc.Util.handleNewInstance(Util.java:425)
--	at com.mysql.jdbc.Util.getInstance(Util.java:408)
--	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:943)
--	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3973)
--	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3909)
--	at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2527)
--	at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2680)
--	at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2497)
--	at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2455)
--	at com.mysql.jdbc.StatementImpl.executeQuery(StatementImpl.java:1369)
--	at sqlancer.common.query.QueryAdapter.executeAndGet(QueryAdapter.java:118)
--	... 11 more
---- Time: 2023/04/18 17:42:35
-- Database: d7
-- Database version: 5.1.0
-- seed value: 1681800583237
```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
main (05b45b531b)

--- a/fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java
+++ b/fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java
@@ -2469,6 +2469,8 @@ public class PlanFragmentBuilder {
                     new PlanFragment(context.getNextFragmentId(), setOperationNode, DataPartition.RANDOM);
             List<List<Expr>> materializedResultExprLists = Lists.newArrayList();
 
+            ScalarOperatorToExpr.FormatterContext formatterContext =
+                    new ScalarOperatorToExpr.FormatterContext(context.getColRefToExpr());
             for (int i = 0; i < optExpr.getInputs().size(); i++) {
                 List<ColumnRefOperator> childOutput = setOperation.getChildOutputColumns().get(i);
                 PlanFragment fragment = visit(optExpr.getInputs().get(i), context);
@@ -2477,8 +2479,7 @@ public class PlanFragmentBuilder {
 
                 // keep output column order
                 for (ColumnRefOperator ref : childOutput) {
-                    SlotDescriptor slotDescriptor = context.getDescTbl().getSlotDesc(new SlotId(ref.getId()));
-                    materializedExpressions.add(new SlotRef(slotDescriptor));
+                    materializedExpressions.add(ScalarOperatorToExpr.buildExecExpression(ref, formatterContext));
                 }
 
                 materializedResultExprLists.add(materializedExpressions);

--- a/fe/fe-core/src/test/java/com/starrocks/analysis/ShowCreateViewStmtTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/analysis/ShowCreateViewStmtTest.java
@@ -12,7 +12,6 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-
 package com.starrocks.analysis;
 
 import com.google.common.collect.Lists;
@@ -167,10 +166,10 @@ public class ShowCreateViewStmtTest {
                 "  |  output exprs:\n" +
                 "  |      VARCHAR | VARCHAR | VARCHAR | VARCHAR\n" +
                 "  |  child exprs:\n" +
-                "  |      VARCHAR | VARCHAR | VARCHAR | VARCHAR\n" +
-                "  |      VARCHAR | VARCHAR | VARCHAR | VARCHAR\n" +
+                "  |      [32: c1, VARCHAR | [33: cast, VARCHAR | [34: cast, VARCHAR | [35: cast, VARCHAR\n" +
+                "  |      [37: c1, VARCHAR | [38: c2, VARCHAR | [42: cast, VARCHAR | [40: c4, VARCHAR\n" +
                 "  |  pass-through-operands: all";
-        Assert.assertTrue(plan.contains(snippet));
+        Assert.assertTrue(plan, plan.contains(snippet));
 
         String dropViewSql = "drop view if exists v2";
         DropTableStmt dropViewStmt = (DropTableStmt) UtFrameUtils.parseStmtWithNewParser(dropViewSql, ctx);

--- a/fe/fe-core/src/test/java/com/starrocks/sql/plan/PlanFragmentWithCostTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/sql/plan/PlanFragmentWithCostTest.java
@@ -730,7 +730,7 @@ public class PlanFragmentWithCostTest extends PlanTestBase {
                 "  2:EXCHANGE\n");
         assertContains(plan, "  STREAM DATA SINK\n" +
                 "    EXCHANGE ID: 02\n" +
-                "    HASH_PARTITIONED: <slot 4>\n" +
+                "    HASH_PARTITIONED: 4: v7\n" +
                 "\n" +
                 "  1:OlapScanNode\n" +
                 "     TABLE: t2");
@@ -894,8 +894,8 @@ public class PlanFragmentWithCostTest extends PlanTestBase {
                     "  |  output exprs:\n" +
                     "  |      [9, BIGINT, true] | [10, BIGINT, true] | [11, BIGINT, true]\n" +
                     "  |  child exprs:\n" +
-                    "  |      [4, BIGINT, true] | [2, BIGINT, true] | [3, BIGINT, true]\n" +
-                    "  |      [8, BIGINT, true] | [6, BIGINT, true] | [7, BIGINT, true]\n" +
+                    "  |      [4: expr, BIGINT, true] | [2: v2, BIGINT, true] | [3: v3, BIGINT, true]\n" +
+                    "  |      [8: expr, BIGINT, true] | [6: v5, BIGINT, true] | [7: v6, BIGINT, true]\n" +
                     "  |  pass-through-operands: all\n");
             assertContains(unionPlan, "  4:OlapScanNode\n" +
                     "     table: t1, rollup: t1\n" +
@@ -929,8 +929,8 @@ public class PlanFragmentWithCostTest extends PlanTestBase {
                     "  |  output exprs:\n" +
                     "  |      [9, BIGINT, true] | [10, BIGINT, true] | [11, BIGINT, true]\n" +
                     "  |  child exprs:\n" +
-                    "  |      [4, BIGINT, true] | [2, BIGINT, true] | [3, BIGINT, true]\n" +
-                    "  |      [8, BIGINT, true] | [6, BIGINT, true] | [7, BIGINT, true]\n");
+                    "  |      [4: expr, BIGINT, true] | [2: v2, BIGINT, true] | [3: v3, BIGINT, true]\n" +
+                    "  |      [8: expr, BIGINT, true] | [6: v5, BIGINT, true] | [7: v6, BIGINT, true]\n");
             Assert.assertTrue(exceptPlan.contains("     probe runtime filters:\n" +
                     "     - filter_id = 0, probe_expr = (5: v4 + 2)"));
             Assert.assertTrue(exceptPlan.contains("     probe runtime filters:\n" +
@@ -943,8 +943,8 @@ public class PlanFragmentWithCostTest extends PlanTestBase {
                     "  |  output exprs:\n" +
                     "  |      [9, BIGINT, true] | [10, BIGINT, true] | [11, BIGINT, true]\n" +
                     "  |  child exprs:\n" +
-                    "  |      [4, BIGINT, true] | [2, BIGINT, true] | [3, BIGINT, true]\n" +
-                    "  |      [8, BIGINT, true] | [6, BIGINT, true] | [7, BIGINT, true]\n");
+                    "  |      [4: expr, BIGINT, true] | [2: v2, BIGINT, true] | [3: v3, BIGINT, true]\n" +
+                    "  |      [8: expr, BIGINT, true] | [6: v5, BIGINT, true] | [7: v6, BIGINT, true]\n");
             Assert.assertTrue(intersectPlan.contains("     probe runtime filters:\n" +
                     "     - filter_id = 0, probe_expr = (5: v4 + 2)"));
             Assert.assertTrue(intersectPlan.contains("     probe runtime filters:\n" +

--- a/fe/fe-core/src/test/java/com/starrocks/sql/plan/ReplayFromDumpTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/sql/plan/ReplayFromDumpTest.java
@@ -239,8 +239,8 @@ public class ReplayFromDumpTest {
                 "  |  output exprs:\n" +
                 "  |      [391, INT, true] | [392, DECIMAL64(7,2), true]\n" +
                 "  |  child exprs:\n" +
-                "  |      [325, INT, true] | [346, DECIMAL64(7,2), true]\n" +
-                "  |      [359, INT, true] | [380, DECIMAL64(7,2), true]"));
+                "  |      [325: ws_sold_date_sk, INT, true] | [346: ws_ext_sales_price, DECIMAL64(7,2), true]\n" +
+                "  |      [359: cs_sold_date_sk, INT, true] | [380: cs_ext_sales_price, DECIMAL64(7,2), true]"));
     }
 
     @Test
@@ -667,8 +667,10 @@ public class ReplayFromDumpTest {
                 "  |  output exprs:\n" +
                 "  |      [16, DATE, false] | [17, BIGINT, true] | [18, DECIMAL128(27,19), true]\n" +
                 "  |  child exprs:\n" +
-                "  |      [2, DATE, false] | [7, BIGINT, true] | [8, DECIMAL128(27,19), true]\n" +
-                "  |      [9, DATE, false] | [14, BIGINT, true] | [15, DECIMAL128(27,19), true]"));
+                "  |      [2: c_0_0, DATE, false] | [7: row_number(), BIGINT, true] " +
+                "| [8: last_value(4: c_0_2), DECIMAL128(27,19), true]\n" +
+                "  |      [9: c_0_0, DATE, false] | [14: row_number(), BIGINT, true] " +
+                "| [15: last_value(11: c_0_2), DECIMAL128(27,19), true]"));
     }
 
     @Test
@@ -690,7 +692,6 @@ public class ReplayFromDumpTest {
                 "  |  column statistics: \n" +
                 "  |  * auto_fill_col-->[1.0, 1.0, 0.0, 1.0, 1.0] ESTIMATE"));
     }
-
 
     @Test
     public void testCorrelatedPredicateRewrite() throws Exception {

--- a/fe/fe-core/src/test/java/com/starrocks/sql/plan/SetTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/sql/plan/SetTest.java
@@ -234,8 +234,8 @@ public class SetTest extends PlanTestBase {
                 "  |  output exprs:\n" +
                 "  |      [26, BIGINT, true] | [27, VARCHAR(20), true] | [28, DOUBLE, true]\n" +
                 "  |  child exprs:\n" +
-                "  |      [1, BIGINT, true] | [4, VARCHAR(20), true] | [5, DOUBLE, true]\n" +
-                "  |      [23, BIGINT, true] | [24, VARCHAR(20), true] | [25, DOUBLE, true]");
+                "  |      [1: v1, BIGINT, true] | [4: cast, VARCHAR(20), true] | [5: cast, DOUBLE, true]\n" +
+                "  |      [23: v4, BIGINT, true] | [24: cast, VARCHAR(20), true] | [25: cast, DOUBLE, true]");
         Assert.assertTrue(plan.contains(
                 "  |  19 <-> [19: k7, VARCHAR, true]\n" +
                         "  |  20 <-> [20: k8, DOUBLE, true]\n" +
@@ -244,12 +244,12 @@ public class SetTest extends PlanTestBase {
         sql = "select * from t0 union all (select cast(v4 as int), v5,v6 " +
                 "from t1 except select cast(v7 as int), v8, v9 from t2)";
         plan = getVerboseExplain(sql);
-        assertContains(plan, "  0:UNION\n" +
+        assertContains(plan, "0:UNION\n" +
                 "  |  output exprs:\n" +
                 "  |      [16, BIGINT, true] | [17, BIGINT, true] | [18, BIGINT, true]\n" +
                 "  |  child exprs:\n" +
-                "  |      [1, BIGINT, true] | [2, BIGINT, true] | [3, BIGINT, true]\n" +
-                "  |      [15, BIGINT, true] | [13, BIGINT, true] | [14, BIGINT, true]\n" +
+                "  |      [1: v1, BIGINT, true] | [2: v2, BIGINT, true] | [3: v3, BIGINT, true]\n" +
+                "  |      [15: cast, BIGINT, true] | [13: v5, BIGINT, true] | [14: v6, BIGINT, true]\n" +
                 "  |  pass-through-operands: all\n" +
                 "  |  cardinality: 2\n" +
                 "  |  \n" +
@@ -276,8 +276,8 @@ public class SetTest extends PlanTestBase {
                 "  |  output exprs:\n" +
                 "  |      [12, INT, true] | [13, BIGINT, true] | [14, BIGINT, true]\n" +
                 "  |  child exprs:\n" +
-                "  |      [7, INT, true] | [5, BIGINT, true] | [6, BIGINT, true]\n" +
-                "  |      [11, INT, true] | [9, BIGINT, true] | [10, BIGINT, true]");
+                "  |      [7: cast, INT, true] | [5: v5, BIGINT, true] | [6: v6, BIGINT, true]\n" +
+                "  |      [11: cast, INT, true] | [9: v8, BIGINT, true] | [10: v9, BIGINT, true]");
     }
 
     @Test
@@ -288,8 +288,8 @@ public class SetTest extends PlanTestBase {
                 "  |  output exprs:\n" +
                 "  |      [5, BOOLEAN, true]\n" +
                 "  |  child exprs:\n" +
-                "  |      [2, BOOLEAN, true]\n" +
-                "  |      [4, BOOLEAN, true]");
+                "  |      [2: expr, BOOLEAN, true]\n" +
+                "  |      [4: expr, BOOLEAN, true]");
 
         sql = "select count(*) from (select 1 as c1 union all select null as c1) t group by t.c1";
         plan = getVerboseExplain(sql);
@@ -297,8 +297,8 @@ public class SetTest extends PlanTestBase {
                 "  |  output exprs:\n" +
                 "  |      [6, TINYINT, true]\n" +
                 "  |  child exprs:\n" +
-                "  |      [2, TINYINT, false]\n" +
-                "  |      [5, TINYINT, true]");
+                "  |      [2: expr, TINYINT, false]\n" +
+                "  |      [5: cast, TINYINT, true]");
 
         sql = "select count(*) from (select cast('1.2' as decimal(10,2)) as c1 union all " +
                 "select cast('1.2' as decimal(10,0)) as c1) t group by t.c1";
@@ -307,8 +307,8 @@ public class SetTest extends PlanTestBase {
                 "  |  output exprs:\n" +
                 "  |      [7, DECIMAL64(12,2), true]\n" +
                 "  |  child exprs:\n" +
-                "  |      [3, DECIMAL64(12,2), false]\n" +
-                "  |      [6, DECIMAL64(12,2), false]\n");
+                "  |      [3: cast, DECIMAL64(12,2), false]\n" +
+                "  |      [6: cast, DECIMAL64(12,2), false]\n");
 
         sql = "select count(*) from (select cast('1.2' as decimal(5,2)) as c1 union all " +
                 "select cast('1.2' as decimal(10,0)) as c1) t group by t.c1";
@@ -317,8 +317,8 @@ public class SetTest extends PlanTestBase {
                 "  |  output exprs:\n" +
                 "  |      [7, DECIMAL64(12,2), true]\n" +
                 "  |  child exprs:\n" +
-                "  |      [3, DECIMAL64(12,2), false]\n" +
-                "  |      [6, DECIMAL64(12,2), false]");
+                "  |      [3: cast, DECIMAL64(12,2), false]\n" +
+                "  |      [6: cast, DECIMAL64(12,2), false]");
     }
 
     @Test
@@ -515,8 +515,8 @@ public class SetTest extends PlanTestBase {
                 "  |  output exprs:\n" +
                 "  |      [9, TINYINT, true]\n" +
                 "  |  child exprs:\n" +
-                "  |      [4, TINYINT, true]\n" +
-                "  |      [8, TINYINT, true]\n" +
+                "  |      [4: day, TINYINT, true]\n" +
+                "  |      [8: day, TINYINT, true]\n" +
                 "  |  pass-through-operands: all");
     }
 

--- a/fe/fe-core/src/test/java/com/starrocks/sql/plan/SetTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/sql/plan/SetTest.java
@@ -12,7 +12,6 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-
 package com.starrocks.sql.plan;
 
 import com.starrocks.common.FeConstants;
@@ -562,7 +561,8 @@ public class SetTest extends PlanTestBase {
         String sql = "set @var = (select v1,v2 from test.t0)";
         StatementBase statementBase = UtFrameUtils.parseStmtWithNewParser(sql, starRocksAssert.getCtx());
         SetExecutor setExecutor = new SetExecutor(connectContext, (SetStmt) statementBase);
-        Assert.assertThrows("Scalar subquery should output one column", SemanticException.class, () -> setExecutor.execute());
+        Assert.assertThrows("Scalar subquery should output one column", SemanticException.class,
+                () -> setExecutor.execute());
         try {
             setExecutor.execute();
             Assert.fail();
@@ -581,5 +581,15 @@ public class SetTest extends PlanTestBase {
                 "  |----4:EXCHANGE\n" +
                 "  |    \n" +
                 "  2:EXCHANGE");
+    }
+
+    @Test
+    public void testUnionNull() throws Exception {
+        String sql = "SELECT DISTINCT NULL\n" +
+                "WHERE NULL\n" +
+                "UNION ALL\n" +
+                "SELECT DISTINCT NULL\n" +
+                "WHERE NULL";
+        getThriftPlan(sql);
     }
 }

